README.md
# simgenphen

`simgenphen` (Simulate Genotypes and Phenotypes) is a wrapper package around several more specialized packages: `bnpsd` for admixture, `simfam` for family structure, and `simtrait` for complex traits.
This package assumes reasonable defaults and generally saves on typing for some recurrent cases.
Key non-trivial functionality is drawing genotypes for admixed families whose loci are all polymorphic.

## Installation

The current development version can be installed from the GitHub repository using `devtools`:
```R
install.packages("devtools") # if needed
library(devtools)
install_github('OchoaLab/simgenphen')
```


## Examples

There are essentially two recommended ways of using this code, shown in the next two subsections.
``` r
library(simgenphen)
```

### One-liner version

Here is a simple way of generating genotypes that feature both admixture and family structure, along with a phenotype!
It's run mostly with default values, except data dimensions are smaller:
```r
data <- sim_gen_phen( n_ind = 50, G = 3, m_loci = 100, m_causal = 5 )

# main objects of interest:
# genotype matrix
data$X
# trait vector
data$trait
```

### Version with separate steps

This approach requires more commands, but is more logical and efficient in loops where population structure is constant but want newly-drawn genotype and phenotype replicates.
These more explicit commands also return additional values.

First construct population parameters (relatedness but no loci):
```r
k_subpops <- 3
data_pop <- sim_pop( n_ind = 50, G = 3, k_subpops = k_subpops )
```
Now draw genotypes!
```r
m_loci <- 100
data_geno <- sim_geno(
    data_pop$admix_proportions_1,
    data_pop$inbr_subpops,
    data_pop$fam,
    data_pop$ids,
    m_loci
)

# genotype matrix
data_geno$X
```
Lastly, draw trait!
In this case we specify a non-default heritability and an environment effect "gcat" to showcase that version (those parameters can also be passed to the "one-liner" function `sim_gen_phen`).
```r
m_causal <- 5
herit <- 0.5
data_trait <- sim_trait_env(
    data_geno$X,
    data_geno$p_anc,
    m_causal,
    herit,
    env = 'gcat',
    k_subpops = k_subpops
)

# the simulated trait
data_trait$trait
# indexes of randomly-selected causal loci
data_trait$causal_indexes
# coefficients of causal loci
data_trait$causal_coeffs
```


NEWS.md
# simgenphen 0.0.0.9000 (2021-08-27)

- First commit of public version
- Includes full tests for all functions
- Nothing exported or documented yet

# simgenphen 0.0.1.9000 (2021-08-27)

- Documented and exported all functions!
- Function `sim_pop` now also returns `admix_proportions_1`, the admixture proportions of the founders, distinguished from `admix_proportions` which is the admixture proportions of the final generation (different if `G > 1`).
- Function `sim_geno`:
  - First parameter is renamed from `admix_proportions` to `admix_proportions_1` to emphasize that it should be the admixture proportions of the founders!
  - Optional parameter `p_anc_input` is renamed to `p_anc`.
- Function `sim_gen_phen` debugged to use founder admixture proportions to draw genotypes for founders.
  The previous version had the bug because `sim_geno` did not return founder admixture proportions, only last-generation values, but the bug only manifested if `G > 1`.
  Bug was very recently introduced (compared to previous non-public code in a separate repository).
- Function `sim_bim` added parameter `pos_gap`.

# simgenphen 0.0.2.9000 (2021-09-03)

- Added package documentation with runnable examples.
- Also updated examples in individual functions to better match the package entry examples.

# simgenphen 0.0.3.9000 (2021-09-03)

- Added README with examples from package documentation.

# simgenphen 0.0.4.9000 (2022-02-21)

- Function `sim_geno` fixed a bug that fixed loci were redrawn without respecting `p_anc` if provided (no bug for `p_anc = NULL`).
- Functions `sim_geno` and `sim_gen_phen` added option `beta` for a Beta distribution for ancestral allele frequencies.

# simgenphen 0.0.5.9000 (2023-02-01)

- Function `sim_gen_phen` added element `p_anc` to return list, needed in order to reuse the simulated genotypes to simulate new traits for them (for example, with different heritabilities).
- Updated unit tests for function `sim_bim` to expect column `alt` before `ref` in return tibble (a change inherited from dependency `genio::make_bim()` made in `genio` version 1.1.0.9000 (2022-04-20))


tests/testthat.R
library(testthat)
library(simgenphen)

test_check("simgenphen")


tests/testthat/test-simgenphen.R
### CHECKERS ###

# write generic checkers that are applied repeatedly

check_kinship <- function( kinship, n_ind ) {
    expect_true( is.matrix( kinship ) )
    expect_equal( nrow( kinship ), n_ind )
    expect_true( isSymmetric( kinship ) )
    expect_true( min( kinship ) >= 0 )
    expect_true( max( kinship ) <= 1 )
}

check_admix_prop <- function( admix, n_ind, k_subpops ) {
    expect_true( is.matrix( admix ) )
    expect_equal( nrow( admix ), n_ind )
    expect_equal( ncol( admix ), k_subpops )
    admix_row_sums <- rowSums( admix ) # calculate
    names( admix_row_sums ) <- NULL # don't want names to be part of comparison
    expect_equal( admix_row_sums, rep.int( 1, n_ind ) )
    expect_true( min( admix ) >= 0 )
    expect_true( max( admix ) <= 1 )
}

check_sim_pop_data <- function ( data, n_ind, k_subpops, fst, G ) {
    # check overall object
    expect_true( is.list( data ) )
    expect_equal( names( data ), c('admix_proportions_1', 'admix_proportions', 'inbr_subpops', 'fam', 'ids', 'kinship') )
    
    # check admixture proportions matrix of founders
    check_admix_prop( data$admix_proportions_1, n_ind, k_subpops )
    
    # check inbr_subpops vector
    inbr_subpops <- data$inbr_subpops
    expect_true( is.numeric( inbr_subpops ) )
    expect_equal( length( inbr_subpops ), k_subpops )
    expect_true( min( inbr_subpops ) >= 0 )
    expect_true( max( inbr_subpops ) <= 1 )
    
    # check kinship of last generation (founders if G=1)
    check_kinship( data$kinship, n_ind )

    # check admixture proportions matrix of last generation (founders if G=1)
    check_admix_prop( data$admix_proportions, n_ind, k_subpops )
    
    # check fst, for admixture params of founders
    admix <- data$admix_proportions_1
    expect_equal( mean( diag( bnpsd::coanc_admix( admix, inbr_subpops ) ) ), fst )
    
    # checks dependent on family structure or not
    if ( G == 1 ) {
        # because there's no family structure here, the last two must be null
        expect_true( is.null( data$fam ) )
        expect_true( is.null( data$ids ) )
    } else {
        # check fam
        fam <- data$fam
        expect_true( is.data.frame( fam ) )
        expect_equal( names( fam ), c('fam', 'id', 'pat', 'mat', 'sex', 'pheno') )
        # number of rows varies because of pruning step, but we do have a clear maximum size
        expect_true( nrow( fam ) <= n_ind * G )
        
        # check ids
        ids <- data$ids
        expect_true( is.list( ids ) )
        expect_equal( length( ids ), G )
    }
}

check_geno <- function( X, n_ind, m_loci ) {
    expect_true( is.matrix( X ) )
    expect_true( all( X %in% c(0L, 1L, 2L) ) )
    expect_equal( nrow( X ), m_loci )
    expect_equal( ncol( X ), n_ind )
    # only special requirement is that no sites are fixed!
    x_bar <- rowMeans( X )
    expect_true( min( x_bar ) > 0 )
    expect_true( max( x_bar ) < 2 )
}

check_p_anc <- function( p_anc, m_loci ) {
    expect_true( is.numeric( p_anc ) )
    expect_equal( length( p_anc ), m_loci )
    expect_true( min( p_anc ) >= 0 )
    expect_true( max( p_anc ) <= 1 )
}

check_sim_geno_data <- function( data, n_ind, m_loci ) {
    # check overall object
    expect_true( is.list( data ) )
    expect_equal( names( data ), c('X', 'p_anc') )

    # check genotypes
    check_geno( data$X, n_ind, m_loci )
    
    # check p_anc
    check_p_anc( data$p_anc, m_loci )
}

check_sim_bim <- function( bim, m_loci, n_chr ) {
    # validate output
    expect_true( is.data.frame( bim ) )
    expect_equal( names( bim ), c('chr', 'id', 'posg', 'pos', 'alt', 'ref') )
    expect_equal( nrow( bim ), m_loci )
    # validate main non-trivial feature, that there are various chromosomes!
    expect_true( all( bim$chr %in% 1 : n_chr ) )
    # ids are unique
    expect_equal( length( bim$id ), length( unique( bim$id ) ) )
    # no positions are zero or negative
    expect_true( all( bim$pos > 0 ) )
    # the rest are defaults from genio::make_bim, meh
}

check_sim_trait_env_data <- function( data, n_ind, m_loci, m_causal ) {
    # check overall object
    expect_true( is.list( data ) )
    expect_equal( names( data ), c('trait', 'causal_indexes', 'causal_coeffs') )

    # check trait
    trait <- data$trait
    expect_true( is.numeric( trait ) )
    expect_equal( length( trait ), n_ind )

    # check causal_indexes
    indexes <- data$causal_indexes
    expect_true( is.numeric( indexes ) )
    expect_equal( length( indexes ), m_causal )
    expect_true( min( indexes ) >= 1 )
    expect_true( max( indexes ) <= m_loci )

    # check causal_coeffs
    coeffs <- data$causal_coeffs
    expect_true( is.numeric( coeffs ) )
    expect_equal( length( coeffs ), m_causal )
}

### TESTS ###

# now actually run specific tests on toy simulated cases

test_that( "sim_pop works", {
    # there are no mandatory params
    
    # default values, but have them defined here to check all dimensions
    n_ind <- 1000
    k_subpops <- 3
    fst <- 0.3
    G <- 1
    expect_silent( 
        data <- sim_pop(
            n_ind = n_ind,
            k_subpops = k_subpops,
            fst = fst,
            G = G,
            verbose = FALSE
        )
    )
    check_sim_pop_data( data, n_ind, k_subpops, fst, G )

    # try higher K
    n_ind <- 1000
    k_subpops <- 10
    fst <- 0.1 # must lower FST to make it all work
    G <- 1
    expect_silent( 
        data <- sim_pop(
            n_ind = n_ind,
            k_subpops = k_subpops,
            fst = fst,
            G = G,
            verbose = FALSE
        )
    )
    check_sim_pop_data( data, n_ind, k_subpops, fst, G )

    # a special case in the code is for high FST
    # (but lower k)
    n_ind <- 1000
    k_subpops <- 3
    fst <- 0.5
    G <- 1
    expect_silent( 
        data <- sim_pop(
            n_ind = n_ind,
            k_subpops = k_subpops,
            fst = fst,
            G = G,
            verbose = FALSE
        )
    )
    check_sim_pop_data( data, n_ind, k_subpops, fst, G )

    # now add generations!
    # return other params to defaults
    n_ind <- 1000
    k_subpops <- 3
    fst <- 0.3
    G <- 3
    expect_silent( 
        data <- sim_pop(
            n_ind = n_ind,
            k_subpops = k_subpops,
            fst = fst,
            G = G,
            verbose = FALSE
        )
    )
    check_sim_pop_data( data, n_ind, k_subpops, fst, G )
    
})

test_that( "sim_geno works", {
    # always have to use sim_pop first, same as before but much smaller sample sizes
    # will check those outputs again here
    n_ind <- 100
    m_loci <- 50
    k_subpops <- 3
    fst <- 0.3
    G <- 1
    p_anc <- NULL
    expect_silent( 
        data <- sim_pop(
            n_ind = n_ind,
            k_subpops = k_subpops,
            fst = fst,
            G = G,
            verbose = FALSE
        )
    )
    check_sim_pop_data( data, n_ind, k_subpops, fst, G )
    admix_proportions_1 <- data$admix_proportions_1
    inbr_subpops <- data$inbr_subpops
    fam <- data$fam
    ids <- data$ids
    kinship <- data$kinship
    # sim_geno has required params without defaults, check that it dies here only (one time only)
    expect_error( sim_geno() )
    expect_error( sim_geno( admix_proportions_1 = admix_proportions_1 ) )
    expect_error( sim_geno( inbr_subpops = inbr_subpops ) )
    # now a successful run
    expect_silent (
        data <- sim_geno(
            admix_proportions_1,
            inbr_subpops,
            fam = fam,
            ids = ids,
            m_loci = m_loci,
            p_anc = p_anc,
            verbose = FALSE
        )
    )
    check_sim_geno_data( data, n_ind, m_loci )

    # change ancestral allele frequencies only
    # this is the only value we've used in practice
    p_anc <- 0.5
    expect_silent (
        data <- sim_geno(
            admix_proportions_1,
            inbr_subpops,
            fam = fam,
            ids = ids,
            m_loci = m_loci,
            p_anc = p_anc,
            verbose = FALSE
        )
    )
    check_sim_geno_data( data, n_ind, m_loci )

    # all the same except with family structure!
    n_ind <- 100
    m_loci <- 50
    k_subpops <- 3
    fst <- 0.3
    G <- 3
    p_anc <- NULL
    expect_silent( 
        data <- sim_pop(
            n_ind = n_ind,
            k_subpops = k_subpops,
            fst = fst,
            G = G,
            verbose = FALSE
        )
    )
    check_sim_pop_data( data, n_ind, k_subpops, fst, G )
    admix_proportions_1 <- data$admix_proportions_1
    inbr_subpops <- data$inbr_subpops
    fam <- data$fam
    ids <- data$ids
    kinship <- data$kinship
    expect_silent (
        data <- sim_geno(
            admix_proportions_1,
            inbr_subpops,
            fam = fam,
            ids = ids,
            m_loci = m_loci,
            p_anc = p_anc,
            verbose = FALSE
        )
    )
    check_sim_geno_data( data, n_ind, m_loci )

    # change ancestral allele frequencies only
    # this is the only value we've used in practice
    p_anc <- 0.5
    expect_silent (
        data <- sim_geno(
            admix_proportions_1,
            inbr_subpops,
            fam = fam,
            ids = ids,
            m_loci = m_loci,
            p_anc = p_anc,
            verbose = FALSE
        )
    )
    check_sim_geno_data( data, n_ind, m_loci )

    # "beta" version
    expect_silent (
        data <- sim_geno(
            admix_proportions_1,
            inbr_subpops,
            fam = fam,
            ids = ids,
            m_loci = m_loci,
            beta = 0.1,
            verbose = FALSE
        )
    )
    check_sim_geno_data( data, n_ind, m_loci )
})

test_that( "sim_bim works", {
    # there are no mandatory parameters
    # pass params for checks
    m_loci <- 100
    n_chr <- 22
    expect_silent( bim <- sim_bim( m_loci, n_chr ) )
    # validate output
    check_sim_bim( bim, m_loci, n_chr )
})

test_that( "sim_trait_env works", {
    # simulate a small X, same as before, but no validations
    # no particular need for family structure or other non-default choices, so keep it simple
    n_ind <- 100
    m_loci <- 50
    k_subpops <- 3
    expect_silent( 
        data <- sim_pop(
            n_ind = n_ind,
            k_subpops = k_subpops,
            verbose = FALSE
        )
    )
    expect_silent (
        data <- sim_geno(
            data$admix_proportions_1,
            data$inbr_subpops,
            m_loci = m_loci,
            verbose = FALSE
        )
    )
    X <- data$X
    p_anc <- data$p_anc
    
    # set other params
    m_causal <- 5
    herit <- 0.5
    env <- NA
    env_var <- 1
    fes <- FALSE
    expect_silent( 
        data <- sim_trait_env(
            X = X,
            p_anc = p_anc,
            m_causal = m_causal,
            herit = herit,
            env = env,
            env_var = env_var,
            fes = fes,
            verbose = FALSE
        )
    )
    check_sim_trait_env_data( data, n_ind, m_loci, m_causal )

    # test FES version
    fes <- TRUE
    expect_silent( 
        data <- sim_trait_env(
            X = X,
            p_anc = p_anc,
            m_causal = m_causal,
            herit = herit,
            env = env,
            env_var = env_var,
            fes = fes,
            verbose = FALSE
        )
    )
    check_sim_trait_env_data( data, n_ind, m_loci, m_causal )

    # test env version, only gcat is implemented
    env <- 'gcat'
    # revert FES (shouldn't matter either way)
    fes <- FALSE
    expect_silent( 
        data <- sim_trait_env(
            X = X,
            p_anc = p_anc,
            m_causal = m_causal,
            herit = herit,
            env = env,
            env_var = env_var,
            k_subpops = k_subpops, # required for `env='gcat'` only
            fes = fes,
            verbose = FALSE
        )
    )
    check_sim_trait_env_data( data, n_ind, m_loci, m_causal )
})

test_that( "sim_gen_phen works", {
    # no parameters are strictly required
    # test a toy case
    n_ind <- 100
    m_loci <- 50
    k_subpops <- 3
    m_causal <- 5
    n_chr <- 22
    # run!
    expect_silent(
        data <- sim_gen_phen(
            n_ind = n_ind,
            m_loci = m_loci,
            k_subpops = k_subpops,
            m_causal = m_causal,
            n_chr = n_chr,
            verbose = FALSE
        )
    )
    # test overall object
    expect_true( is.list( data ) )
    expect_equal( names( data ), c('X', 'bim', 'p_anc', 'kinship', 'admix_proportions', 'trait', 'causal_indexes', 'causal_coeffs') )
    # check X
    check_geno( data$X, n_ind, m_loci )
    # check bim
    check_sim_bim( data$bim, m_loci, n_chr )
    # check p_anc
    check_p_anc( data$p_anc, m_loci )
    # check kinship
    check_kinship( data$kinship, n_ind )
    # check admixture proportions matrix
    check_admix_prop( data$admix_proportions, n_ind, k_subpops )
    # check trait components, simple subset of given data
    check_sim_trait_env_data(
        data[ c('trait', 'causal_indexes', 'causal_coeffs') ],
        n_ind, m_loci, m_causal
    )

    # "beta" version
    expect_silent(
        data <- sim_gen_phen(
            n_ind = n_ind,
            m_loci = m_loci,
            k_subpops = k_subpops,
            m_causal = m_causal,
            n_chr = n_chr,
            beta = 0.1,
            verbose = FALSE
        )
    )
    # test overall object
    expect_true( is.list( data ) )
    expect_equal( names( data ), c('X', 'bim', 'p_anc', 'kinship', 'admix_proportions', 'trait', 'causal_indexes', 'causal_coeffs') )
    # check X
    check_geno( data$X, n_ind, m_loci )
    # check bim
    check_sim_bim( data$bim, m_loci, n_chr )
    # check p_anc
    check_p_anc( data$p_anc, m_loci )
    # check kinship
    check_kinship( data$kinship, n_ind )
    # check admixture proportions matrix
    check_admix_prop( data$admix_proportions, n_ind, k_subpops )
    # check trait components, simple subset of given data
    check_sim_trait_env_data(
        data[ c('trait', 'causal_indexes', 'causal_coeffs') ],
        n_ind, m_loci, m_causal
    )
})


R/sim_bim.R
#' Construct a variant table in plink BIM format with several chromosomes.
#'
#' Purpose is to build an artificial BIM table with several chromosomes, which some methods require, such as leave-one-chromosome-out (LOCO) estimation.
#' This function produces chromosomes with approximately equal numbers of variants.
#' Positions are spaced uniformly by a given size.
#' 
#' @param pos_gap The number of basepairs to space variants.
#' @inheritParams sim_gen_phen
#'
#' @return A `tibble` with these columns:
#' - `chr`: The constructed chromosome positions.
#' - `id`: Locus ID values (autocompleted by [genio::make_bim()] to match row numbers).
#' - `posg`: Genetic distance (autocompleted by [genio::make_bim()] to `0`, all missing).
#' - `pos`: locus positions, starting at `pos_gap` for each chromosome and increasing in increments of `pos_gap`.
#' - `ref`: Reference allele (autocompleted by [genio::make_bim()] to `1`).
#' - `alt`: Alternative allele (autocompleted by [genio::make_bim()] to `2`).
#'
#' @examples
#' # simulate a small table
#' m_loci <- 100
#' bim <- sim_bim( m_loci )
#' 
#' @export
sim_bim <- function(
                    m_loci = 100000,
                    n_chr = 22,
                    pos_gap = 1000
                    ) {
    bim <- tibble::tibble(
        # chromosomes are all roughly equal length in this case
        chr = ceiling( (1 : m_loci ) / m_loci * n_chr ),
        pos = 0 # initialize for now
    )
    # add positions in this loop, contiguous per chr, restarting after each case
    for ( chr in 1 : n_chr ) {
        indexes <- which( bim$chr == chr )
        bim$pos[ indexes ] <- ( 1 : length( indexes ) ) * pos_gap # space them
    }
    # autocomplete the rest
    bim <- genio::make_bim( bim )
    # return that tibble
    return( bim )
}


R/sim_pop.R
#' Simulate population structure
#'
#' This function defines admixture and optionally family parameters establishing how individuals are related to each other.
#' Admixture proportions are deterministically constructed from the linear 1D admixture model [bnpsd::admix_prop_1d_linear()].
#' If `fst < 0.5` the intermediate inbreeding coefficients are proportional to their index (`1:k_subpops`), otherwise they are proportional to `(1:k_subpops) + 1`.
#' If `G > 1`, a random family structure is drawn using [simfam::sim_pedigree()].
#' No per-locus parameters are constructed or simulated with this function.
#'
#' @inheritParams sim_gen_phen
#'
#' @return A list containing the following elements:
#' - `admix_proportions_1`: The true admixture proportions of the founders (required to draw genotypes for admixed founders and seed the family structure if `G > 1`)
#' - `admix_proportions`: The true admixture proportions of the joint admixed family model (final generation only).
#' - `inbr_subpops`: The vector of intermediate subpopulation inbreeding/FST values for admixture model.
#' - `fam`: If `G > 1`, the pedigree structure as a plink FAM table, pruned with [simfam::prune_fam()] to include only ancestors of the final generation; otherwise `NULL`.
#' - `ids`: If `G > 1`, a list of IDs from `fam` above split into non-overlapping generations; otherwise `NULL`.
#' - `kinship`: The true kinship of the joint admixed family model (final generation only).
#'
#' @examples
#' # a small example
#' data_pop <- sim_pop( n_ind = 50, G = 3 )
#' 
#' # parameters for admixture model of founders
#' data_pop$admix_proportions_1
#' data_pop$inbr_subpops
#' 
#' # parameters of family structure
#' data_pop$fam
#' data_pop$ids
#' 
#' # parameters for final generation
#' data_pop$admix_proportions
#' data_pop$kinship
#' 
#' @seealso
#' The `bnpsd` and `simfam` packages.
#'
#' @export
sim_pop <- function(
                    n_ind = 1000,
                    k_subpops = 3,
                    bias_coeff = 0.5,
                    fst = 0.3,
                    G = 1,
                    verbose = TRUE
                    ) {
    # FST values for admixture subpopulations
    inbr_subpops <- 1 : k_subpops
    # high admixture FSTs require subpopulations to all be more differentiated, this works when k==3
    if ( fst >= 0.5 )
        inbr_subpops <- inbr_subpops + 1
    
    # define population structure
    if (verbose)
        message('admix_prop_1d_linear')
    obj <- bnpsd::admix_prop_1d_linear(
        n_ind = n_ind,
        k_subpops = k_subpops,
        bias_coeff = bias_coeff,
        coanc_subpops = inbr_subpops,
        fst = fst
    )
    # in this case return value is a named list with three items:
    admix_proportions_1 <- obj$admix_proportions # admixture proportions
    inbr_subpops <- obj$coanc_subpops # rescaled Fst vector for intermediate subpops

    # get true kinship matrix
    if (verbose)
        message('coanc_admix, coanc_to_kinship')
    coancestry <- bnpsd::coanc_admix( admix_proportions_1, inbr_subpops ) # the coancestry matrix
    kinship <- bnpsd::coanc_to_kinship( coancestry ) # kinship matrix
    
    # should be null when there's no family structure
    fam <- NULL
    ids <- NULL
    # let these be the final ones either way (if G == 1 then use *_1)
    admix_proportions_G <- admix_proportions_1
    if (G > 1) {
        # simulate realistic generations of families

        # simulate pedigree first
        if (verbose)
            message('sim_pedigree')
        data_simfam <- simfam::sim_pedigree( n_ind, G )
        fam <- data_simfam$fam # the pedigree itself
        ids <- data_simfam$ids # to filter generations later

        # prune fam table now, to not simulate unnecessary individuals without descendants
        if (verbose)
            message('prune_fam')
        fam <- simfam::prune_fam( fam, ids[[ G ]] )

        if (verbose)
            message('kinship_last_gen')
        # label founders in matrix
        rownames( kinship ) <- ids[[ 1 ]]
        colnames( kinship ) <- ids[[ 1 ]]
        # now calculate total kinship in final generation
        kinship <- simfam::kinship_last_gen( kinship, fam, ids )

        if (verbose)
            message('admix_last_gen')
        # label founders in matrix
        rownames( admix_proportions_1 ) <- ids[[ 1 ]]
        # get correct admixture proportions for these children!
        admix_proportions_G <- simfam::admix_last_gen( admix_proportions_1, fam, ids )
    }
    
    return(
        list(
            admix_proportions_1 = admix_proportions_1,
            admix_proportions = admix_proportions_G,
            inbr_subpops = inbr_subpops,
            fam = fam,
            ids = ids,
            kinship = kinship
        )
    )
}



R/sim_trait_env.R
#' Simulate a complex trait
#'
#' Mostly a wrapper around [simtrait::sim_trait()], but can optionally add a structured environment effect.
#' However, in that case the heritability is no longer as specified.
#'
#' @param X The genotype matrix.
#' @param p_anc The vector of ancestral allele frequencies.
#' @param k_subpops The number of intermediate subpopulations for admixture model.
#' Used for `env = 'gcat'` only.
#' @inheritParams sim_gen_phen
#'
#' @return A list containing the following elements:
#' - `trait`: The simulated trait vector.
#' - `causal_indexes`: The vector of randomly selected causal loci indexes.
#' - `causal_coeffs`: The vector of simulated regression coefficients for causal loci.
#'
#' @examples
#' # draw population parameters using `sim_pop` first
#' # a small example
#' k_subpops <- 3
#' data_pop <- sim_pop( n_ind = 50, G = 3, k_subpops = k_subpops )
#'
#' # then draw genotypes
#' m_loci <- 100
#' data_geno <- sim_geno(
#'     data_pop$admix_proportions_1,
#'     data_pop$inbr_subpops,
#'     data_pop$fam,
#'     data_pop$ids
#' )
#'
#' # now draw trait!
#' m_causal <- 5
#' herit <- 0.5
#' data_trait <- sim_trait_env(
#'     data_geno$X,
#'     data_geno$p_anc,
#'     m_causal,
#'     herit,
#'     env = 'gcat',
#'     k_subpops = k_subpops
#' )
#'
#' # the simulated trait
#' data_trait$trait
#' # indexes of randomly-selected causal loci
#' data_trait$causal_indexes
#' # coefficients of causal loci
#' data_trait$causal_coeffs
#'
#' @export
sim_trait_env <- function(
                          X,
                          p_anc,
                          m_causal,
                          herit,
                          env = NA,
                          env_var = 1,
                          k_subpops = NA,
                          fes = FALSE,
                          verbose = TRUE
                          ) {

    if (verbose) message('simtrait')

    # parameters of simulation

    # create simulated trait and associated data
    # use version for known p_anc (prefered, only applicable to simulated data)
    # let sim_trait complain if its params are missing
    obj <- simtrait::sim_trait(
        X = X,
        m_causal = m_causal,
        herit = herit,
        p_anc = p_anc,
        fes = fes
    )
    
    if (!is.na(env) && env == 'gcat') {
        if ( is.na( k_subpops ) )
            stop( '`k_subpops` is required if `env = "gcat"`!' )
        
        # construct non-genetic environment effect
        # this is similar, in spirit, to the GCAT paper simulations
        n_ind <- length( obj$trait )
        # in this simulation, cluster membership is given by existing index:
        # the number is used as the environmental effect
        env <- ceiling( ( 1 : n_ind ) / n_ind * k_subpops )
        # rescale with sample variance, to have unit variance
        #        env <- env / sqrt( var(env) )
        # center and scale
        env <- as.numeric( scale( env ) )
        # scale to get desired variance 
        env <- env * sqrt( env_var )
        # simple combine
        obj$trait <- obj$trait + env
    }

    # return what we want from trait simulation
    return(
        list(
            trait = obj$trait,
            causal_indexes = obj$causal_indexes,
            causal_coeffs = obj$causal_coeffs
        )
    )
}


R/sim_gen_phen.R
#' Simulate genotypes and phenotypes
#'
#' Simulate entire parameters and random variables, particularly random genotypes and phenotypes, with desired dimensions and other base parameters.
#'
#' @param n_ind The number of individuals.
#' @param m_loci The number of loci.
#' @param k_subpops The number of intermediate subpopulations for admixture model.
#' @param fst The FST of the admixed individuals (for the founders if `G > 1`).
#' @param bias_coeff The bias coefficient of the admixed individuals (for the founders if `G > 1`).
#' @param G The number of generations for random family.
#' The `G = 1` case is no family structure (just admixture).
#' For `G > 1` first admixed founders are simulated, then the family structure is simulated from them.
#' @param m_causal The number of causal loci for the trait, selected randomly from among the simulated loci.
#' @param herit The trait heritability.
#' @param env A string describing environment model.
#' Only `NA` (no environment) or "gcta" are accepted.
#' @param env_var The variance of the environment effect.
#' Ignored if `env` is `NA`.
#' @param fes If `FALSE` (default) constructs trait from Random Coefficients (RC) model.
#' If `TRUE`, the Fixed Effect Sizes (FES) model is used instead.
#' @param n_chr Number of chromosomes to simulate.
#' Chromosome assignments are not biologically meaningful, as all loci are drawn independently (no LD).
#' @param beta Shape parameter for a symmetric Beta for ancestral allele frequencies `p_anc`.
#' If `NA` (default), `p_anc` is uniform with range in \[0.01, 0.5\].
#' Otherwise, `p_anc` has a symmetric Beta distribution with range in \[0, 1\].
#' @param verbose If `TRUE` reports progress, otherwise it is silent.
#'
#' @return A list containing the following elements:
#' - `X`: The simulated genotype matrix.
#' - `bim`: The variant info table.
#' - `p_anc`: The ancestral allele frequency vector (aligned with rows of `X`).
#' - `kinship`: The true kinship of the joint admixed family model (final generation only).
#' - `admix_proportions`: The true admixture proportions of the joint admixed family model (final generation only).
#' - `trait`: The simulated trait vector.
#' - `causal_indexes`: The vector of randomly selected causal loci indexes.
#' - `causal_coeffs`: The vector of simulated regression coefficients for causal loci.
#'
#' @examples
#' # run with default values, except smaller
#' data <- sim_gen_phen( n_ind = 50, G = 3, m_loci = 100, m_causal = 5 )
#'
#' # main objects of interest:
#' # genotype matrix
#' data$X
#' # trait vector
#' data$trait
#'
#' @seealso
#' This function is a wrapper around [sim_pop()], [sim_geno()], [sim_bim()], and [sim_trait_env()], see those for more details.
#'
#' @export
sim_gen_phen <- function(
                         # params of population structure
                         n_ind = 1000, # number of individuals (columns)
                         m_loci = 100000, # number of loci (rows)
                         k_subpops = 3, # number of intermediate subpopulations
                         fst = 0.3, # desired final Fst
                         bias_coeff = 0.5, # bias coeff of standard Fst estimator
                         # the family code gets triggered if G > 1
                         G = 1, # number of generations (1 is unrelated founders)
                         # params of complex trait
                         m_causal = 100,
                         herit = 0.8,
                         env = NA,
                         env_var = 1,
                         fes = FALSE,
                         n_chr = 22,
                         beta = NA,
                         verbose = TRUE # to print messages
                         ) {
    # this function is for generating random genotypes from an admixed population with 3 source subpopulations
    # meant to abstractly resemble Hispanics and African-Americans

    # things that can't be changed:
    # - Fst of intermediate subpopulations is a ramp proportional to 1:k
    # - Admixture proportions are from 1D geography model

    ##################
    ### 0: SIM POP ###
    ##################

    obj <- sim_pop(
        n_ind = n_ind,
        k_subpops = k_subpops,
        bias_coeff = bias_coeff,
        fst = fst,
        G = G,
        verbose = verbose
    )
    admix_proportions_1 <- obj$admix_proportions_1
    admix_proportions <- obj$admix_proportions
    inbr_subpops <- obj$inbr_subpops
    fam <- obj$fam
    ids <- obj$ids
    kinship <- obj$kinship
    
    ####################
    ### 1: GENOTYPES ###
    ####################

    obj <- sim_geno(
        admix_proportions_1 = admix_proportions_1,
        inbr_subpops = inbr_subpops,
        fam = fam,
        ids = ids,
        m_loci = m_loci,
        beta = beta,
        verbose = verbose
    )
    X <- obj$X
    p_anc <- obj$p_anc

    # BOLT requires more than one chr, this aims for human chrs (vaguely)
    bim <- sim_bim(
        m_loci = m_loci,
        n_chr = n_chr
    )
    
    ################
    ### 2: TRAIT ###
    ################

    obj <- sim_trait_env(
        X = X,
        p_anc = p_anc,
        m_causal = m_causal,
        herit = herit,
        env = env,
        env_var = env_var,
        k_subpops = k_subpops,
        fes = fes,
        verbose = verbose
    )
    trait <- obj$trait
    causal_indexes <- obj$causal_indexes
    causal_coeffs <- obj$causal_coeffs

    #################
    ### 3: RETURN ###
    #################

    # return a few items of interest
    # NOTES:
    # - inbr_subpops not used outside
    # - could return fam/ids, don't right now
    return(
        list(
            X = X, # genotype matrix
            bim = bim,
            p_anc = p_anc,
            kinship = kinship, # kinship matrix of last generation (founders if G=1)
            admix_proportions = admix_proportions, # admixture proportions of last generation (founders if G=1)
            trait = trait, # trait vector
            causal_indexes = causal_indexes, # randomly-picked causal locus index
            causal_coeffs = causal_coeffs # locus effect size vector (for causal loci only, rest are zero)
        )
    )
}
